### DISCUSSION TITLE
token数获取

### DISCUSSION THREAD

**Discussion by 5478-lfy** [lobehub/lobehub#1365] [Development Question] [answered]
![image](https://github.com/lobehub/lobe-chat/assets/92673213/06dfef0b-d6c1-44a4-ac60-65664342501f)
我想要在响应结束之后获取这个值，可以吗，如何获取

**Comment by 5478-lfy** [comment]
我想要代理获取可以吗，这是我的openresty代理nginx配置
#user  nobody;
worker_processes  1;

#error_log  logs/error.log;
#error_log  logs/error.log  notice;
#error_log  logs/error.log  info;

#pid        logs/nginx.pid;


events {
    worker_connections  1024;
}


http {
    include       mime.types;
        default_type  application/octet-stream;

    log_format  main '$remote_addr - $remote_user [$time_local] "$request" '
                      '$status $body_bytes_sent "$http_referer" '
                      '"$http_user_agent" "$http_x_forwarded_for"';

    access_log  logs/access.log  main;


    # client_body_in_file_only off;


    sendfile        on;
    #tcp_nopush     on;

    keepalive_timeout  65;

    gzip  on;

    upstream sd_service{
        server 127.0.0.1:7862; # 这里是后端服务1的IP和端口
        # 可以添加多个后端服务器，Nginx会根据配置的策略进行负载均衡
    }

    upstream comfyui_srevice{
        server 127.0.0.1:8188; # 这里是后端服务2的IP和端口
    }

    upstream face_srevice{
        server 127.0.0.1:7863; # 这里是后端服务2的IP和端口
    }

    upstream jupyter_srevice{
        server 127.0.0.1:7863; # 这里是后端服务2的IP和端口
    }





    server {
        listen       443 ssl;
        server_name  aiex.mosslab.hq.com;

	#SSL-START SSL相关配置，请勿删除或修改下一行带注释的404规则
        ssl_certificate      /etc/openresty/ssl/aiex.mosslab.hq.com.pem;
        ssl_certificate_key  /etc/openresty/ssl/aiex.mosslab.hq.com.key;
        ssl_session_cache shared:SSL:1m;
        ssl_session_timeout 5m;
        #TLS协议版本越高，HTTPS通信的安全性越高，但是相较于低版本TLS协议，高版本TLS协议对浏览器的兼容性较差。
        ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE:ECDH:AES:HIGH:!NULL:!aNULL:!MD5:!ADH:!RC4;
        ssl_protocols TLSv1.1 TLSv1.2 TLSv1.3;


        location /api/chat/openai {
            proxy_pass http://127.0.0.1:3210;
            proxy_set_header Cookie $http_cookie; # 确保Cookie头被转发到后端服务
            access_by_lua_block {
            if ngx.var.request_method == "POST" then
                ngx.req.read_body()
                local data = ngx.req.get_body_data()
                -- 如果请求体数据为nil，则设置为JSON中的null
                if not body_data then
                    ngx.log(ngx.ERR, "Request body data is nil")
                    body_data = ' "" '
                end
                -- 获取时间戳
                local timestamp = ngx.time()
                local time = os.date("%Y-%m-%d %H:%M:%S")

                -- 获取特定的 Cookie 值
                local cookie_dingdinginfo = ngx.var.cookie_dingdinginfo or ' "" '
                local decoded_dingdinginfo = ngx.unescape_uri(cookie_dingdinginfo)

                local cookie_prompt_tokens = ngx.var.cookie_prompt_tokens or ' "" '
                local cookie_total_tokens = ngx.var.cookie_total_tokens or ' "" '
                
                -- 获取当前日期，用于生成日志文件名
                local current_date = os.date("%Y-%m-%d")
                local file_path = string.format("/var/log/openresty/%s.log", current_date)



                -- 构建JSON格式的字符串
                local request_info = string.format(
                    '{"chat_info": %s,"user_info": {"dingdinginfo": %s}, "prompt_tokens": "%s", "total_tokens": "%s","timestamps": "%s"}',
                    data,
                    decoded_dingdinginfo,
                    ngx.escape_uri(cookie_prompt_tokens),
                    ngx.escape_uri(cookie_total_tokens),
                    time
                )

                -- 保存到日志文件
                local file = io.open(file_path, "a")
                if file then
                    file:write(request_info .. "\n")
                    file:close()
                else
                    ngx.log(ngx.ERR, "Failed to open file: " .. file_path)
                end
            end
        }

    }






        
    #ssl_prefer_server_ciphers  on;
	location /sd/ {
            proxy_pass http://127.0.0.1:7862; # 请求/service1/路径的请求会被转发到8080端口的服务
	    access_log  /var/log/openresty/aiex.mosslab.hq.com.sd.access.log  main;
        }

	location /jupyter/ {
            proxy_pass http://jupyter_srevice; # 请求/service1/路径的请求会被转发到8080端口的服务
        }

	location /comfyui/ {
            proxy_pass http://comfyui_srevice; # 请求/service1/路径的请求会被转发到8080端口的服务
        }

	location /face/ {
            proxy_pass http://face_srevice; # 请求/service1/路径的请求会被转发到8080端口的服务
        }

    location /v1.0/ {
        proxy_pass https://api.dingtalk.com;
    }
    location /topapi/ {
        proxy_pass https://oapi.dingtalk.com;
    }
    
	location / {
		proxy_pass http://127.0.0.1:3210/;
        proxy_buffering off;
		proxy_set_header Host $host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                access_log  /var/log/openresty/aiex.mosslab.hq.com.ask.access.log  main;
                       
        }
        access_log  /var/log/openresty/aiex.mosslab.hq.com.access.log  main;
	    error_log   /var/log/openresty/aiex.mosslab.hq.com.error.log;
    }

}

**Comment by 5478-lfy** [comment]
或者说响应里面有哪个参数是存储这个tokens的吗

**Reply by arvinxx** [reply, marked as answer]
目前没有。这个数据是前端实时计算的
